src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -2670,11 +2675,17 @@ public int hashCode() {
 
         @Override
         public boolean equals(Object val) {
-            return val instanceof CSS.LengthValue lu && span == lu.span;
+            if (percentage) {
+                return false;
+            } else {
```

Comment:
Why are percentage values never equal?

I expect two attribute sets produced from `"margin-top: 100%"` to be equal. However, the real length on rendering could be different if the attribute is applied to different elements.

Reply:
I was not sure how to get the other value which is told in the spec where it says percentage needs another value to be passed in. Also checking for percentage equality resulted in regression in one closed test bug4690812.java so percentage equality was not handled.

https://github.com/openjdk/jdk/blob/8aa50288a1a6a853bf4d8d97b1849d5744ed7a32/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2607-L2609

Reply:
> I was not sure how to get the other value which is told in the spec where it says percentage needs another value to be passed in. Also checking for percentage equality resulted in regression in one closed test bug4690812.java so percentage equality was not handled.
> 
> https://github.com/openjdk/jdk/blob/8aa50288a1a6a853bf4d8d97b1849d5744ed7a32/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2607-L2609

I don't think you need to. In this case, AttributeSets are compared. All the relative units will result in different computed value depending on the element they're applied to.

I'll take a look at the test. I can't see how implementation of `equals` affects that test.

Reply:
> I don't think you need to. In this case, AttributeSets are compared. All the relative units will result in different computed value depending on the element they're applied to.

Here's a [sample HTML file](https://cr.openjdk.org/~aivanov/7083187/percentage-values.html). Both `<p>` and `<span>` element have CSS rule `font-size: 200%`. The computed value for `p` is based on its parent font size, which is the `body` element. The computed value of `span` is also based on its parent font size, its parent is `p` element, therefore the font-size of `span` is twice as large as in `p` and four times larger compared to `body`.

At the same time, the attribute sets applied to `p` and `span` should be considered equal because they contain the same rules.